### PR TITLE
DAOS-13379 object: fixes for object upgrade

### PR DESCRIPTION
**Pull request by wangdi1**
Upgrade id layout version for non EC object.

migrate specific shard for non-EC object upgrade.

Delete the old layout version object once upgrade
finished.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by wangshilong — DISMISSED** [review]
I manually verified upgrade from 2.2 to master with this PR, it fixed problem.

File: src/vos/vos_obj.c [hunk 1]
```diff
@@ -621,8 +621,8 @@ vos_obj_key2anchor(daos_handle_t coh, daos_unit_oid_t oid, daos_key_t *dkey, dao
 	return rc;
 }
 
-int
-vos_obj_delete(daos_handle_t coh, daos_unit_oid_t oid)
+static int
+_vos_obj_delete(daos_handle_t coh, daos_unit_oid_t oid, bool only_delete_entry)
```
Comment (jolivier23) [inline src/vos/vos_obj.c:625]: this is fine but generally we use _internal or _helper in other places

File: src/vos/vos_obj_index.c [hunk 2]
```diff
@@ -111,30 +118,47 @@ oi_rec_alloc(struct btr_instance *tins, d_iov_t *key_iov,
...
-			DP_RC(rc));
-		return rc;
-	}
+	/* Normally it should delete both ilog and vo_tree, but during upgrade
+	 * the new OID (with new layout version) will share the same ilog and
+	 * vos_tree with the old OID (with old layout version), so it will only
+	 * delete the entry in this case.
+	 */
+	if (del_arg != NULL && del_arg->only_delete_entry) {
+		memset(&obj->vo_ilog, 0, sizeof(obj->vo_ilog));
```
Comment (jolivier23) [inline src/vos/vos_obj_index.c:144]: I'm confused by this change, I will take up up offline.   But first comment is that the ilog may allocate new space internally so memset could result in storage leak.

Comment (wangdi1) [reply src/vos/vos_obj_index.c:144]: During upgrade,  the new OI entry does not need to recreate the ilog or vos_tree, instead they will share the same ilog and vos_tree with the old entry, and these old entry fee will only reset the ilog and vos_tree.

**Review by jolivier23 — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Review by wangshilong — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Comment by wangdi1** [comment]
failure is due to DAOS-13191.

**Pull request merged**